---
layout: page
title: 2019 Project ideas list
permalink: /project-ideas/2019/
---

The 2018 project ideas list is still available here: [2018 project ideas list](http://www.intermine.org/gsoc/project-ideas/2018)

## Ideas list: 

### Upgrade im-tables and imjs

**Description**: [imjs](https://github.com/intermine/imjs) is a node.js library for accessing InterMine APIs, and [im-tables](https://github.com/intermine/im-tables) is a browser-based data table displayer for InterMine data that uses imjs to communicate with the server. Both are written in CoffeeScript. Your task would be to update the dependencies (last seriously updated in 2015) and add additional upgrades and bugfixes for these libraries. This would include adding methods to communicate with the [InterMine registry](http://registry.intermine.org/), adding displayers to the [column summary](https://flymine.readthedocs.io/en/latest/results-tables/Documentationresultstables.html#column-summaries) features, and migrating the im-tables tests to use mocked communication rather than the InterMine Test Model. 

**Required skills:**

- JavaScript

**Possible mentors:** 

- Yo Yehudi
- Aman Dwivedi
- Adrián Rodríguez-Bazaga (if needed)

You can find the mentors in the [InterMine chat](http://chat.intermine.org). 


**Expected outcome:** Updated versions of im-tables and imjs with the new features.
**How could this integrate with existing InterMine tools?** This could be rolled out immediately on all JSP pages for InterMine Server / the legacy InterMine UI. 
**Difficulty level:** Hard if student is unfamiliar with coffeescript, medium otherwise. 


### Javascript Data Visualisations

**Title**: Create a suite of javascript-based biological data visualisations. These visualisations would be designed to work with [BlueGenes](https://github.com/intermine/bluegenes), InterMine's ClojureScript based user interface, and would need to conform to the [Bluegenes Tool API](https://intermineorg.wordpress.com/2018/12/03/javascript-everywhere-the-bluegenes-tool-api-version-1-is-released/). Clojure knowledge is not required. 

**Required skills:**

- CSS
- HTML
- Javascript
- DOM manipulation

**Possible mentors:** 

- Yo Yehudi. 
- Aman Dwivedi
- Adrián Rodríguez-Bazaga (if needed)

You can find the mentors in the [InterMine chat](http://chat.intermine.org). 


**Expected outcome:** A suite of 2-5 javascript based tools that can be used in production BlueGenes.
**How could this integrate with existing InterMine tools?** This will be used directly within BlueGenes. 
**Difficulty level:** Medium if student is familiar with javascript visualisation techniques, Hard if not. 

---------------

### Dataset Versioning tool

**Description**: InterMine integrates data from many different sources. We have [existing Perl scripts](https://github.com/intermine/intermine-scripts/tree/master/bio/DataDownloader/lib/DataDownloader/Source) that download the data, which we would like to extend to automatically scrape the version details from each data source. This information would be added to the InterMine data model, and be made available via webservice. 

See [project requirements](https://docs.google.com/document/d/1MjjWSNbYFhvuyP0w0ibn7J9C5ik35y6ODS-s-l6tWZI/edit) for details.

**Required skills:**

- Java
- Perl 

**Possible mentors:** 

- Julie Sullivan
- Rachel Lyne

You can find the mentors in the [InterMine chat](http://chat.intermine.org). 


**Expected outcome:** Updated InterMine code that makes recording the data set version automatic. 
**How could this integrate with existing InterMine tools?** This would be used directly by InterMine Server. 
**Difficulty level:** Easy, slightly harder if you aren't familiar with Perl or Java. 

-------------------------
### InterMine data modeller tool

**Description**: When someone adds a data type to their InterMine, their first step is to look around at other InterMine instances to see how everyone else modelled that data type. With so many InterMines (over 30!) this is hard to do. We'd like a tool that can automate this. For example, this tool should answer the questions: "Does anyone have 'pathway' data in their InterMine? If so, how did they model it?"

See [project requirements](https://docs.google.com/document/d/1xisOjf3vG37A4WPXt-IMdWumUegd6TnQXVmj7AlmVXo/edit#) for details.

**Required skills:**

- Language skills - you'll need to know one of: JavaScript, Python, or another [common language from the InterMine stack](https://github.com/intermine/). 

**Possible mentors:** 

- Julie Sullivan
- Rachel Lyne

You can find the mentors in the [InterMine chat](http://chat.intermine.org). 

**Expected outcome:** A production ready tool that interrogates the data models of InterMines, and presents the information to users in an easy to digest way.
**How could this integrate with existing InterMine tools?** This would probably be a standalone tool that is intended for InterMine admins, or could become part of an InterMine Admin Portal. 
**Difficulty level:** Easy

-------------------------
### InterMine Schema Validator

**Description**: There are many different biological file formats in use, and people do not always comply with the standards correctly, which can make parsing the files a challenge. We would like to create a modular library that accepts certain file formats as input and reports the validity of the file, with details of any errors encountered. 

See [project requirements](https://docs.google.com/document/d/1yNxyGI7lWN7CoJFo5SHiBryIzBMOZaieT_nKU56Qerw/edit) for details.

**Required skills:**

- Java

**Possible mentors:** 

- Julie Sullivan
- Yo Yehudi
- Adrián Rodríguez-Bazaga (if needed)


You can find the mentors in the [InterMine chat](http://chat.intermine.org). 

**Expected outcome:** A library that will validate biological data file formats like [GFF](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md) and [FASTA](https://en.wikipedia.org/wiki/FASTA_format), designed in a modular manner so further file formats can be added.  
**How could this integrate with existing InterMine tools?** This would be a standalone library that can be used by InterMine as a dependency. 
**Difficulty level:** Easy.

-------------------------
### Migrate RESTful web services from Struts to Spring

**Description**: [InterMine](https://github.com/intermine/intermine) was created in 2002, and uses an [end-of-life version of Struts](https://struts.apache.org/index.html) which is no longer supported. InterMine provides RESTful web services to access the integrated data programmatically: they allow to execute custom or templated queries, search keywords, manage lists, discover metadata, perform enrichment statistics and manage user profiles. We would like to migrate all [InterMine RESTful webservices](http://iodocs.apps.intermine.org/) to [Spring](https://spring.io/) and document the APIs with Swagger, compliant with the [Open API Specifications](https://github.com/OAI/OpenAPI-Specification).
There is a robust test suite in-place which should make it easier to migrate without introducing errors, and students are encouraged to add tests to any areas lacking coverage. 

**Required skills:**

- Java
- Spring
- Web services

**Possible mentors:** 

- Daniela Butano
- Julie Sullivan

You can find the mentors in the [InterMine chat](http://chat.intermine.org). 


**Expected outcome:** Spring integrated with the [existing InterMine webservices](https://github.com/intermine/intermine), with some or all of the webservices migrated from Struts to Spring, and tests passing for all migrated services. APIs documented with Swagger.
**How could this integrate with existing InterMine tools?** This will be part of InterMine Server. 
**Difficulty level:** Medium, but could be easier if a student is familiar with Spring. 

-------------------------
### InterMineR, a client library for R

**Description**: InterMine is an open source biological data warehouse. InterMine has several client libraries which allow users to write scripts to access InterMine data directly via the command line. We have an R client that is popular but would like to add more features.

See [project requirements](https://docs.google.com/document/d/1lwf_yijv8cyRXt6HSJF6oMzo0rRuxdM-L_3xC2MySeM/edit#) for details.


**Required skills:**

- Good understanding of R
- Familiarity with [Bioconductor](https://www.bioconductor.org/)

**Possible mentors:** 

- Rachel Lyne
- Julie Sullivan

You can find the mentors in the [InterMine chat](http://chat.intermine.org). 


**Expected outcome:** An upgraded InterMineR client that is released on [BioConductor](https://www.bioconductor.org/)
**How could this integrate with existing InterMine tools?** This client will be used to query the InterMine API. 
**Difficulty level:** Medium.

-------------------------
### InterMine Similarity Finder Tool

**Description**: InterMine instances contain large amounts of biological data. We would like to provide an explorative interface that would allow scientists to visualise and explore different ways that genes (or other biological entites) might be related. This interface would be highly configurable and probably based on [cytoscape.js](http://js.cytoscape.org/). 

**Required skills:**

- Javascript
- Browser skills, e.g. DOM manipulation, CSS, UX / HCI


**Possible mentors:** 

- Yo Yehudi
- Rachel Lyne

You can find the mentors in the [InterMine chat](http://chat.intermine.org). 

**Expected outcome:** A javascript-powered user interfacethat directly integrates into [BlueGenes](https://github.com/intermine/bluegenes) using the [BlueGenes tool API](https://intermineorg.wordpress.com/2018/12/03/javascript-everywhere-the-bluegenes-tool-api-version-1-is-released/)
**How could this integrate with existing InterMine tools?** This would be a standalone javascript package, directly embedded in BlueGenes.
**Difficulty level:** Medium, or easier if the student is already familiar with Cytoscape.js.

-------------------------

### Unit test suite for Clojure application

**Description**: BlueGenes ([GitHub](https://github.com/intermine/bluegenes), [Demo](http://bluegenes.apps.intermine.org/)) is a [ClojureScript](https://clojurescript.org/) browser-based UI for [InterMine](http://intermine.org/) that doesn't have many tests. Currently, before merging a pull request, reviewers need to manually validate certain pages work correctly using a [checklist](https://github.com/intermine/bluegenes/blob/dev/PULL_REQUEST_TEMPLATE.md). Ideally many of these checks could be automated into a test suite to save reviewers time and making it harder to miss errors! See our possible [testcase list](https://docs.google.com/spreadsheets/d/1R6JZcHMoyXCFgP7Qm-UH4Jzr-rNqBPunBHxiTWrfs2I/edit#gid=0) (this list is still growing!). More info on the project available [here](https://docs.google.com/document/d/1x6cFwXHVVKrKblKpmh3PrPobWHJuD1qsPvEPhgg4si4/edit#). 


**Required skills:**

- Clojurescript / Clojure
- Advantageous: ability to examine software and identify+implement useful test cases.

**Possible mentors:** 

- Yo Yehudi
- Adrián Rodríguez-Bazaga


You can find the mentors in the [InterMine chat](http://chat.intermine.org).

**Expected outcome:** A UI test suite using a mock server for http requests.  This could be using cypress, phantomjs or similar.
**How could this integrate with existing InterMine tools?** e.g. BlueGenes, InterMine Server.
**Difficulty level:** Hard, or medium if students know both Clojure and have written tests before.

-------------------------

### Improving the InterMine Data Browser

**Description**: The InterMine Data Browser ([GitHub](https://github.com/intermine/InterMine-Data-Browser-Tool), [Demo](http://data-browser.apps.intermine.org/)) is a JavaScript browser-based tool that aggregates and presents data across InterMine instances. The objective of this tool is to alleviate the heterogeneity produced by the usage of totally different data platforms, by exposing a user-friendly interface on top of InterMine data services.

The goal for this project, is to extend the Data Browser with new features. The new functionalities to be implemented are open to the student innovation capabilities. As an example, one new feature could be to add new filters, such as the [Phenotypes filter](https://github.com/intermine/InterMine-Data-Browser-Tool/issues/21). Other examples and problems to be fixed are available at the [issues page](https://github.com/intermine/InterMine-Data-Browser-Tool/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc).


**Required skills:**

- JavaScript
- NodeJS
- CSS
- Some knowledge of API requests handling can be advantageous

**Possible mentors:** 

- Adrián Rodríguez-Bazaga
- Second mentor name here


You can find the mentors in the [InterMine chat](http://chat.intermine.org).

**Expected outcome:** A new [InterMine Data Browser](https://github.com/intermine/InterMine-Data-Browser-Tool) release where new features and functionalities are implemented.

**How could this integrate with existing InterMine tools?** It is directly integrated as it extends the InterMine Data Browser.

**Difficulty level:** Medium if the student knows JavaScript and NodeJS, Hard otherwise.

